Port elastictube1d to tutorials

carme-hp opened the pull request:
Motivation for this pull request is to adapt the elastictube1d case to make it coherent with the new planned structure for the tutorials. 

An orientative to-do list is added here:

- [x] Case structure should follow https://github.com/precice/tutorials/issues/139
- [x] Single precice-config.xml
- [x] cxx case: same solver for the parallel run (use flag "-parallel")
- [x] Change naming according to convention (eg. "-" instead "_")

carme-hp commented:
**Updated and more detailed list of what to do**

- [x] Single precice config file
- [x] update all-clean files
- [x] update all-run files
- [x] python tools folder with tubePlot, thetaScheme, output, configurationFile
- [x] src directory (with headers) for cxx? 
- [x] update included headers for cxx
- [x] cxx parallel run must be implemented as a flag
- [x] renaming according to other tutorials

carme-hp commented:
**Single FluidSolver.cpp for both serial and parallel run**
@MakisH I have uploaded a file named FluidSolverFlag.cpp with my idea of how to have a single file and to use the tag "-parallel". I need to fix further stuff in order to check if this is working but I would like to know if this is the way to go:)

MakisH commented:
Thank you! Before I look into this, could you please describe what you expect the user to run for each case? Is this currently depicted in the help messae of the solver and in the README?

carme-hp commented:
at the moment I haven't written any information on that but the one in the README.md is still valid

> Thank you! Before I look into this, could you please describe what you expect the user to run for each case? Is this currently depicted in the help messae of the solver and in the README?

carme-hp commented:
Do we want to keep Allrun scripts? If so how should they fit in the case structure suggested in https://github.com/precice/tutorials/issues/139?

MakisH commented:
> Do we want to keep Allrun scripts? If so how should they fit in the case structure suggested in [precice/tutorials#139](https://github.com/precice/tutorials/issues/139)?

Not really, we can remove the `Allrun` scripts. We don't even use them in the system tests anymore and it is easy to reconstruct them if we have good `run.sh` scripts.

carme-hp commented:
**Update on the elastictube1d case:**

There is one single `precice-config.xml` to be used for both python and cpp versions. The versions can be combined as wished:) BUT, there is an issue:
The results using fluid-cpp or python-cpp differ. To reproduce the issue run both options in serial and plot the diameter. In case of python the diameter at the end of the tube is smaller than at the start, what this does not happen in the cpp solution. I'm not sure what makes more sense physically.

@MakisH  I  plotted the diameter at the end of the tube but it's not clear to me how to interpret the graph. For the watch-point I used (It was not clear to me if this is the coordinate I should write):
```
 <watch-point mesh="Fluid-Nodes" name="outlet" coordinate="10.0; 0.0"/>
```





![watchpoint](https://user-images.githubusercontent.com/71499004/110981160-0b563000-8367-11eb-9877-ac24f334acaf.png)

MakisH commented on README.md:
```diff
@@ -1,112 +1,114 @@
+## Setup
 
-- [preCICE](https://github.com/precice/precice/wiki/Get-preCICE)
+An accurate descrption of the setup and the physics of the problem is described [here](https://github.com/precice/precice/wiki/1D-elastic-tube:-Case-Description). 
```
We need to update all the references to the wiki to the new website.
Also, try to avoid links with "here" as a title. Instead, write something like "in the documentation page X".
But in any case, we probably should move all the information into this page.

carme-hp replied:
i have written below "ALTERNATIVE TO THE LINK" what I think it would make sense to have here

MakisH replied:
The tutorial on the website will be here: https://www.precice.org/tutorials-elastic-tube-1d.html

MakisH commented on fluid-cpp/fluid_solver.cpp:
```diff
@@ -0,0 +1,173 @@
+
+#include "precice/SolverInterface.hpp"
+#include <iostream>
+#include <mpi.h>
```
Interestingly, with GCC 10.2 and OpenMPI 4.0.3, I get the following warning:
```
[ 25%] Building CXX object CMakeFiles/FluidSolver.dir/fluid_solver.cpp.o
In file included from /usr/lib/x86_64-linux-gnu/openmpi/include/openmpi/ompi/mpi/cxx/mpicxx.h:277,
                 from /usr/lib/x86_64-linux-gnu/openmpi/include/mpi.h:2868,
                 from /home/makish/github/elastictube1d/fluid-cpp/fluid_solver.cpp:6:
/usr/lib/x86_64-linux-gnu/openmpi/include/openmpi/ompi/mpi/cxx/op_inln.h: In member function ‘virtual void MPI::Op::Init(void (*)(const void*, void*, int, const MPI::Datatype&), bool)’:
/usr/lib/x86_64-linux-gnu/openmpi/include/openmpi/ompi/mpi/cxx/op_inln.h:121:25: warning: cast between incompatible function types from ‘void (*)(void*, void*, int*, ompi_datatype_t**, void (*)(void*, void*, int*, ompi_datatype_t**))’ to ‘void (*)(void*, void*, int*, ompi_datatype_t**)’ [-Wcast-function-type]
  121 |     (void)MPI_Op_create((MPI_User_function*) ompi_mpi_cxx_op_intercept,
      |                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/lib/x86_64-linux-gnu/openmpi/include/openmpi/ompi/mpi/cxx/op_inln.h:123:38: warning: cast between incompatible function types from ‘void (*)(const void*, void*, int, const MPI::Datatype&)’ to ‘void (*)(void*, void*, int*, ompi_datatype_t**)’ [-Wcast-function-type]
  123 |     ompi_op_set_cxx_callback(mpi_op, (MPI_User_function*) func);
      |                                      ^~~~~~~~~~~~~~~~~~~~~~~~~
```
Since this warning has (I understand) nothing to do with our code, we should suppress it.

@fsimonis we had a similar issue in preCICE, right? How did we solve it there?

fsimonis replied:
I cannot remember this warning in preCICE, but it doesn't look like an issue we have to deal with. Let's ignore it.

MakisH commented on fluid-cpp/fluidComputeSolution.cpp:
```diff
@@ -276,24 +279,217 @@ void fluidComputeSolution(
+    double tau)
+{
+  /* fluid_nl Variables */
+  int i, j = 0, k, ampl;
```
Warning for unused variable:
```
/home/makish/github/elastictube1d/fluid-cpp/fluidComputeSolution.cpp:322:20: warning: variable ‘ampl’ set but not used [-Wunused-but-set-variable]
  322 |   int i, j = 0, k, ampl;
      |                    ^~~~
```

MakisH commented on README.md:
```diff
@@ -1,112 +1,114 @@
+A working known combination for the input parameters is N=100, tau = 0.01, kappa = 100. Other parameters like the simulation's end time, you can modify them in the precice-config.xml.
 
-### Requirements
+Note that you first need to build the scripts `FluidSolver` and `SolidSolver`. Each script needs to be build separately.
```
Better move this before the "running" part.

MakisH commented on README.md:
````diff
@@ -1,112 +1,114 @@
+```
+python3 ./solid-python/SolidSolver.py precice-config.xml 
+```
+Parameters such as N can be modified directly at the `FluidSolver.py` and at the `SolidSolver.py`. The parameters must be coherent between the different solvers and participants. 
````
```suggestion
Parameters such as N can be modified directly at the `FluidSolver.py` and at the `SolidSolver.py`. The parameters must be consistent between the different solvers and participants. 
```

MakisH commented on README.md:
````diff
@@ -1,112 +1,114 @@
-$ python Postproc/fluid.py diameter Postproc/out_fluid_
+$ python3 Postproc/fluid.py diameter Postproc/out_fluid_
+```
+![FSI3 setup](images/diameter.png)
````
I guess that the image needs to be updated, right?

carme-hp replied:
I keep this open in case the updated image is not the final one:)

MakisH commented on Allclean:
```diff
@@ -4,15 +4,15 @@
+      Solid.log
 
 rm -r precice-run/
 
```
Missing one file:

```suggestion
rm -rfv precice-Solid-events-summary.log
```

The `-v` (verbose) option also displays when a file is deleted, the `-f` (force) does not complain if the file is not there.

carme-hp replied:
sorry, the -rfv flag should be used for the other files or only this missing one?

MakisH replied:
It should be also for the rest. But don't worry too much about the `Allclean`, I will anyway try to make a general one.

carme-hp replied:
okay thanks!

MakisH commented on Allclean:
```diff
@@ -4,15 +4,15 @@
       precice-STRUCTUR-events-summary.log \
       Postproc/*.vtk \
       Fluid.log \
-      Structure.log
+      Solid.log
```
I understand that these files are not there anymore, right? Or are they from the Python side?

carme-hp replied:
that's true, I have removed them now:)

MakisH commented on README.md:
```diff
@@ -1,112 +1,114 @@
+Both fluid and solid participant are supported in:
 
-**Note:** these requirements are specific to the Python variant of the code. If you wish to run the C++ version, you require different packages (see [this section](#c-version)).
+* *cxx*
+* *python*: 
```
Is something missing here?

MakisH commented on fluid-python/FluidSolver.py:
```diff
@@ -2,10 +2,10 @@
 from thetaScheme import perform_partitioned_implicit_trapezoidal_rule_step, perform_partitioned_implicit_euler_step
 import numpy as np
-import tubePlotting
+import tubePlotting as tubePlotting
```
Is this really needed?

MakisH reviewed (changes requested):
Here are some first comments, mainly on the `README.md`. I will continue with the review tonight.

MakisH commented on .gitignore:
```diff
@@ -0,0 +1,12 @@
+solid-cpp/build/
+fluid-python/__pycache__/
+solid-python/__pycache__/
+convergence-*
+iterations-*
+.scon*
```
Not needed anymore:
```suggestion
```

MakisH commented on Postproc/fluid.py:
```diff
@@ -7,7 +7,7 @@
 from mpl_toolkits.mplot3d import Axes3D
 
-T = 100  # number of timesteps performed
+T = 99  # number of timesteps performed
```
Why this change?

MakisH commented on README.md:
```diff
@@ -1,112 +1,137 @@
+title: 1D Elastic Tube
+permalink: unknown
+keywords: OpenFOAM, python
+summary: The 1D Elastic Tube is a FSI case, that consists of an internal flow in a flexible tube. The flow is unsteady and incompressible. This tutorial can be run using C++ or python.  Running the simulation takes a few minutes. 
```
```suggestion
summary: The 1D Elastic Tube is a FSI case, that consists of an internal flow in a flexible tube. The flow is unsteady and incompressible. This tutorial contains C++ and Python variants of the fluid and structure solvers.  Running the simulation takes just 1-2 minutes. 
```

MakisH commented on README.md:
```diff
@@ -1,112 +1,137 @@
+![FSI3 setup](images/setup.png)
 
-### Requirements
+Fluid is assumed to be incompressible flow and gravity is neglected. Due to the axisymmetry, the flow can be described using a quasi-two dimensional continuity and momentum equations. The motivation and exact formulation of the equations that we consider can be found in [2]. 
```
```suggestion
The flow is assumed to be incompressible and gravity is neglected. Due to the axisymmetry, the flow can be described using a quasi two dimensional continuity and momentum equations. The motivation and exact formulation of the equations that we consider can be found in [2]. 
```
(I am very confused if it should be `quasi-two-dimensional`, `quasi two dimensional`, `quasi-two dimensional`, or `quasi two-dimensional`... :confused: )

carme-hp replied:
I found results with every combination but since I see that Nature uses `quasi-two-dimensional` I guess I can take this one as the good one.

MakisH commented on README.md:
```diff
@@ -1,112 +1,137 @@
-- [preCICE](https://github.com/precice/precice/wiki/Get-preCICE)
+The following parameters have been chosen:
+- Length of the tube: L = 10
+- Inlet velocity: v_inlet = 10 + 3 sin (10 PI t)
```
I think LaTeX also works:
```suggestion
- Inlet velocity: $$ v_inlet = 10 + 3 sin (10 PI t) $$
```
See https://www.precice.org/docs-meta-cheatsheet.html#latex-math-syntax

carme-hp replied:
actually, the latex thingie it's not working for me, you can see the $$ in the readme...

MakisH replied:
I just hope it will appear correctly in the website. If not, we change it back later.

MakisH commented on README.md:
```diff
@@ -1,112 +1,137 @@
+- Young modulus: E = 10000
 
-- [preCICE Python Bindings](https://github.com/precice/python-bindings).
+LINK:
```
todo

MakisH commented on README.md:
````diff
@@ -1,112 +1,137 @@
-    ```bash
-    $ cd python/ && ./Allrun
-    ```
+The python version is realized using the Python API for preCICE. Check the [preCICE wiki](https://github.com/precice/precice/wiki/1D-elastic-tube-using-the-Python-API) for more information on this example.
````
todo: link update

MakisH commented on README.md:
````diff
@@ -1,112 +1,137 @@
+```
+mpiexec -np <nproc> ./solid-cpp/build/SolidSolver precice-config.xml N -parallel
+```
+A working known combination for the input parameters is N=100, tau = 0.01, kappa = 100. Other parameters like the simulation's end time, you can modify them in the precice-config.xml.
````
```suggestion
A working known combination for the input parameters is N=100, tau = 0.01, kappa = 100.
```
I think we can safely remove this line, as this is anyway explained in the Quickstart and other places.

MakisH commented on README.md:
````diff
@@ -1,112 +1,137 @@
-   ```bash
-   $ make all
-   ```
+### python
````
Let's use `Python` with uppercase `P` everywhere in the `README.md`, for consistency with the rest of the documentation.

MakisH commented on README.md:
```diff
@@ -1,112 +1,137 @@
+
+## Post-processing
+
+The `Postproc/` folder contains the .vtk files resulting from the execution, which you can visualize using eg. paraview. Alternatively you can visualize the results with the provided `Postproc/fluid.py` script:
```
```suggestion
The `Postproc/` folder contains the .vtk files resulting from the execution, which you can visualize using e.g. ParaView. Alternatively you can visualize the results with the provided `Postproc/fluid.py` script:
```

(sorry for being so pedantic, but these are also easy fixes :) )

MakisH commented on README.md:
````diff
@@ -1,112 +1,137 @@
-$ python Postproc/fluid.py <quantity> Postproc/<prefix>
+$ python3 Postproc/fluid.py <quantity> Postproc/<prefix>
 ```
 Note the required arguments specifying which quantity to plot (`pressure`, `velocity` or `diameter`) and a name prefix for the target vtk files.
````
```suggestion
Note the required arguments specifying which quantity to plot (`pressure`, `velocity` or `diameter`) and the name prefix of the target vtk files.
```

For a moment I was confused that I needed to select a prefix to assign.

MakisH commented on README.md:
````diff
@@ -1,112 +1,137 @@
+$ python3 Postproc/fluid.py diameter Postproc/out_fluid_
+```
+
+An image of this diameter plot can be found in the `/images` folder.
````
Which file are you referring to exactly? Let's refer to it directly (at least by filename).

carme-hp replied:
actually i think it makes sense to delete the whole sentence

MakisH commented on README.md:
````diff
@@ -1,112 +1,137 @@
+$ python3 Postproc/fluid.py diameter Postproc/out_fluid_
 ```
-An image of this diameter plot can be found in the `cxx/example` folder.
+![FSI3 setup](images/python.png)
````
Please also rename the images to fit the pattern described in https://github.com/precice/tutorials/pull/163.
The reason is that the website system (Jekyll) collects all the image files into one directory, so we are doing some "namespacing" by ourselves.

carme-hp replied:
Then I understand I should use `tutorials-elastictube1d-diameter.png` and `tutorials-elastictube1d-setup.png` ?

MakisH replied:
Yes, exactly.

MakisH commented on README.md:
````diff
@@ -1,112 +1,137 @@
+$ python3 Postproc/fluid.py diameter Postproc/out_fluid_
+```
+
+An image of this diameter plot can be found in the `/images` folder.
````
Maybe we could also enable the watchpoint in the `precice-config.xml` and add the picture here? The picture is already in `images/`.

We could also delete the old `images/diameter.png`.

MakisH commented on fluid-cpp/fluidComputeSolution.cpp:
```diff
@@ -1,8 +1,10 @@
-#include "FluidSolver.h"
+#include "FluidComputeSolution.h"
```
You already changed the solvers and this was a very much needed work! :smiley:  Could you please go through the file diff and self-review the most important changes in the solvers? Add some comments on why some changes were needed (e.g. why removing the `_n` variables? What else should I pay attention at?): this will help you realize if there are any more bugs and it would help me review the PR.

Files I have in mind:
- [ ] `fluid-cpp/fluidComputeSolution.cpp`
- [ ] `fluid-cpp/fluidSolver.cpp` (especially since there is no diff after the renaming)
- [x] `fluid-python/fluidSolver.py`
- [x] `solid-cpp/solidComputeSolution.cpp`
- [x] `solid-cpp/solidSolver.cpp`
- [x] `solid-python/solidSolver.py`

MakisH replied:
I looked over the solid and all the python solvers, I only need some help in reviewing the fluid C++ solver changes.

MakisH reviewed (changes requested):
A few more comments. Going through the solvers next, it would be helpful if you could also self-review there (see my comment with specific files).

MakisH commented on solid-cpp/Solid_solver.cpp:
```diff
@@ -0,0 +1,136 @@
+#include <iostream>
+#include <stdlib.h>
+#include "mpi.h"
+#include "SolidSolver.h"
```
I see a naming inconsistency here:
- `SolidSolver.h`
- `Solid_solver.cpp`
- `SolidComputeSolution.cpp`
- `fluid_solver.cpp`

I guess that they should be:
- `fluid_solver.cpp` and `solid_solver.cpp`, or `FluidSolver.cpp` and `SolidSolver.cpp`
- the respective headers should be `fluid_solver.h` and `solid_solver.h`, or `FluidSolver.h` and `SolidSolver.h`
- not sure if we really need the `...ComputeSolution.cpp` files to be separate.

Of course, we should not rewrite all the internals here, but I think the file names are important for this PR.

carme-hp replied:
Good point! I decided to go for the capital option (aka `FluidSolver.cpp`)

MakisH commented on fluid-python/FluidSolver.py:
```diff
@@ -100,18 +113,19 @@
-    velocity, pressure, success = perform_partitioned_implicit_euler_step(velocity_n, pressure_n, crossSectionLength_n,
-                                                                          crossSectionLength, dx, precice_dt, config.velocity_in(t + precice_dt), custom_coupling=False)
+    velocity, pressure, success = perform_partitioned_implicit_euler_step(velocity_old, pressure_old, crossSectionLength_old,
+                                                                          crossSectionLength, dx, precice_dt, velocity_in(t + precice_dt), custom_coupling=True)
```
Why the False/True change?

carme-hp replied:
I changed that to have the same we do in the cpp-solver. In any case, it didn't notice any difference in the results (at least not an obvious one).

MakisH commented on fluid-python/output.py:
```diff
@@ -0,0 +1,77 @@
+    n_datasets = data.__len__()
+    assert n_datasets == datanames.__len__()
+
+    outpath = os.path.join(os.getcwd(), 'Postproc')
```
I would prefer to name all directories starting with a lowercase, actually (for consistency with the rest, and for easy guessing/tab-completion):
```suggestion
    outpath = os.path.join(os.getcwd(), 'postproc')
```
(similarly, update the path and any other scripts)

MakisH commented on fluid-cpp/fluid_solver.cpp:
```diff
@@ -0,0 +1,173 @@
+  std::cout << "Starting Fluid Solver..." << std::endl;
+  if (argc != 5 && argc != 6) {
+    std::cout << std::endl;
+    std::cout << "Fluid: Usage: mpiexec -np <#procs> " << argv[0] << " <configurationFileName> <N> <tau> <kappa> -parallel" << std::endl;
```
Since this is always tedious to lookup, let's add simple `run.sh` scripts, similarly to the tutorials. Here they will probably not be longer than five lines.

MakisH commented on precice-config.xml:
```diff
@@ -0,0 +1,72 @@
+         defined. The name of the participant has to match the name given on construction of the
+         precice::SolverInterface object used by the participant. -->
+    
+    <participant name="FLUID">
```
```suggestion
    <participant name="Fluid">
```

(similarly everywhere else)

MakisH commented on precice-config.xml:
```diff
@@ -0,0 +1,72 @@
+
+    <coupling-scheme:serial-implicit>
+      <participants first="FLUID" second="Solid"/>
+      <max-time value="0.99"/>
```
Why not a perfect `1`? Looks suspicious...

carme-hp replied:
Changed! That was necessary at some point but not anymore:)

MakisH commented on precice-config.xml:
```diff
@@ -0,0 +1,72 @@
+    <data:scalar name="CrossSectionLength"/>
+
+    <!-- A common mesh that uses these data fields -->
+    <mesh name="Fluid-Nodes">
```
The meshes should now be names `Fluid-Mesh-Nodes` etc (we want the `-Mesh` part there, for consistency).

MakisH commented on solid-python/SolidSolver.py:
```diff
@@ -4,12 +4,25 @@
+p0 = 0  # pressure at outlet
+L = 10  # length of tube/simulation domain
+E = 10000  # elasticity module
+c_mk = np.sqrt(E/2/r0)  # wave speed
```
This appears everywhere as a derived quantity, could you maybe try printing its value and comparing?

carme-hp replied:
i checked and its same:)

MakisH reviewed (changes requested):
A few more comments, the last for tonight. :-)

MakisH commented:
A few suggestions also from @uekerman regarding the different results:

1. The most probably source of trouble is the inlet velocity profile. And with a quick look, I see indeed a small difference that I do not understand.

In the [C++ version](https://github.com/precice/elastictube1d/blob/b3a38fcfd73edb82c36661105456ddfaef526730/fluid-cpp/FluidComputeSolution.cpp#L382-L384):
```
    /* Velocity Inlet is prescribed */
    tmp1 = 10 + 3* sin(10* PI * (t+0.01)); //inlet velocity
    Res[0] = tmp1 - velocity[0];
```

In the [Python version](https://github.com/precice/elastictube1d/blob/b3a38fcfd73edb82c36661105456ddfaef526730/fluid-python/FluidSolver.py#L55):
```
velocity_in = lambda t: u0 + ampl * np.sin(frequency * (t+t_shift) * np.pi)  # inflow velocity
```
where `u0 = 10`,`frequency = 10`, `ampl = 3`, and `t_shift = 0`. Why is `t_shift = 0.01` in the C++? This is small, anyway.

2. Another potential pitfall can be due to the dimensionless timestepping: we have a given end time and we ask for a different number of time steps. Do these aggree?

3. He also suspects a `+`/`-` typo somewhere, which I know that is very painful to explore and that you have already invested a lot of time on this.

4. Suggestions to move on: In case of doubt, we keep the Python version as the "truth". The parallel version we could even drop, it does not add much value. If we don't manage to fix this, let's drop the C++ all along. I would suggest here to include the "broken" C++ and add a disclaimer in the README that there is an open issue regarding this (which we need to open and link to).

carme-hp commented:
>     1. The most probably source of trouble is the inlet velocity profile. And with a quick look, I see indeed a small difference that I do not understand.

This doesn't make a noticeable difference in the results but I removed it from the cpp for consistency. 

>     1. Another potential pitfall can be due to the dimensionless timestepping: we have a given end time and we ask for a different number of time steps. Do these aggree?

This is a good point, but I printed the timing for python and I can see it increases by 0.01 in every time step so I guess this is not the issue. 

>     3. Suggestions to move on: In case of doubt, we keep the Python version as the "truth". The parallel version we could even drop, it does not add much value. If we don't manage to fix this, let's drop the C++ all along. I would suggest here to include the "broken" C++ and add a disclaimer in the README that there is an open issue regarding this (which we need to open and link to).

I would like to try to find whats wrong until Friday and if I can't then I will just drop the parallel version and we can write the C++ as an open issue.

carme-hp commented:
# C++ results are not consistent with python

The results obtained with the python and c++ versions differ.  The python version is the one believed to be correct. The difference in the results originates from the fluid solver (fluid-cpp and solid-cpp lead to the same results than fluid-cpp and solid-python) . The difference in the results can be observed by naked eye after plotting with 
```
python3 postproc/fluid.py diameter postproc/cpp_fluid_
```
![cpp](https://user-images.githubusercontent.com/71499004/114310578-78d2b900-9aeb-11eb-994d-8da0bd021674.png)
for the cpp results and 
```
python3 postproc/fluid.py diameter postproc/out_fluid_
```
for the python results. 
![python](https://user-images.githubusercontent.com/71499004/114310579-796b4f80-9aeb-11eb-9c58-7b6752e3917e.png)
You can see that for the python tube the diameter is smaller at the end of the tube. This is the expected behavior, as we are further away from the inlet. 

I have tried to debug this issue unsuccessfully so far. For anyone who wants to try: 
- `alpha` and `dx` don't have the same value in each version. `alpha=0` in python but `alpha=50` in cpp, whereas `dx=0.1` in python but `dx=0.001` in cpp .
- If we use the python's values of `alpha` and `dx`, then the cpp does not converge. If we print values we see that the first two iterations look like in python but then pressure explodes.  
- The most likely explanation for that would be some typo in a sign but I couldn't find it. 

Finally, if someone manages to fix this issue, please make sure to adress the [parallel issue](https://github.com/precice/elastictube1d/pull/39#issuecomment-817318591) as well!

carme-hp commented:
# Parallel run

First, note that there is an open issue regarding the [cpp results](https://github.com/precice/elastictube1d/pull/39#issuecomment-817318591). Second, note that ´FluidSolver´ calls different functions in case of a serial or a parallel run. This is of course not desirable and both functions should be unified when the mentioned issue is fixed.

For now, the function `fluidComputeSolutionParallel` is solving a different scenario than `fluidComputeSolutionSerial`. The scenario that `fluidComputeSolutionParallel` is solving is the first one that was historically developed for elastic tube and is different than the one that was implemented in python later on. I modified `fluidComputeSolutionSerial` a posteriori to mimic the scenario solved in the python version and this is the reason why we have to different scenarios in the cpp.

In any case, this is what the results when using `fluidComputeSolutionSerial`version looked like originally:

![diameter](https://user-images.githubusercontent.com/71499004/114310822-7de43800-9aec-11eb-9a83-462a85f36b09.png)

If you do a parallel run you will be able to reproduce this results. But because the domain is now split into different processors you will obtain a fraction of this plot.  For example we can visualize the results for `rank = 2` by 

```
python3 postproc/fluid.py diameter postproc/out_fluid2_
```

MakisH commented:
Thank you for documenting! Could you please open separate issues for each? https://github.com/precice/elastictube1d/issues

MakisH commented:
TODO now (I can do these):
- [x] Rename `postproc/fluid.py` to something more descriptive (e.g. `plot-fluid.py` or something better) and update the `README.md`.
- [x] Formal the preCICE config file, remove comments, remove explicit default values.
- [x] Do we need to update or to remove this statement? `An accurate description of the setup and the physics of the problem is described [here](https://github.com/precice/precice/wiki/1D-elastic-tube:-Case-Description). `
- [x] Same for `The python version is realized using the Python API for preCICE. Check the [preCICE wiki](https://github.com/precice/precice/wiki/1D-elastic-tube-using-the-Python-API) for more information on this example.`
- [x] Replace `run.sh` with the latest state.
- [x] Somehow I don't get result files when running at least the C++ version
- [x] Do we still need something to replace `ALTERNATIVE TO THE LINK:`?
- [x] Shall we add [this animation ](https://raw.githubusercontent.com/wiki/precice/precice/images/tube.gif) at the end?
- [x] Compare with the wiki pages
- [x] The Python results are now the same as C++ (but in the bad way)

TODO when moving to `tutorials/`:
- Update `clean.sh` to use the common tools.
- Comment on removing the exact equations and refering to the paper
- Remove the old wiki pages and link to website

carme-hp commented:
> Same for The python version is realized using the Python API for preCICE. Check the [preCICE wiki](https://github.com/precice/precice/wiki/1D-elastic-tube-using-the-Python-API) for more information on this example.

i could directly refer to https://github.com/precice/python-bindings and mention libraries on my readme, if you think that makes more sense.

carme-hp commented:
> Do we need to update or to remove this statement? An accurate description of the setup and the physics of the problem is described [here](https://github.com/precice/precice/wiki/1D-elastic-tube:-Case-Description). 

I think it is safe to remove this

MakisH reviewed (approved):
Ready to merge! :tada:
